Authoring Gradle Builds section in Gradle User Manual (#26188)
gradle/gradle: subprojects/docs/src/docs/userguide/authoring-builds/introduction/writing_tasks.adoc

= Writing Tasks

Gradle tasks are created by extending link:{groovyDslPath}/org.gradle.api.DefaultTask.html[`DefaultTask`].

However, the generic `DefaultTask` provides no action for Gradle.
If users want to extend the capabilities of Gradle and their build script, they must either use a *built-in task* or create a *custom task*:

1. *Built-in task* - Gradle provides built-in utility tasks such as `Copy`, `Jar`, `Zip`, `Delete`, etc...
2. *Custom task* - Gradle allows users to subclass `DefaultTask` to create their own task types.

[[sec:sample_task]]
== Create a task

The simplest and quickest way to create a *custom* task is in a build script:

To create a task, inherit from the `DefaultTask` class and implement a `@TaskAction` handler:

.build.gradle.kts
[source,kotlin]
----
abstract class CreateFileTask : DefaultTask() {
    @TaskAction
    fun action() {
        val file = File("myfile.txt")
        file.createNewFile()
        file.writeText("HELLO FROM MY TASK")
    }
}
----

The `CreateFileTask` implements a simple set of *actions*.
First a file called "myfile.txt" is created in the main project.
Then, some text is written to the file.

== Register a task

> A task is *registered* in the build script using the `register()` method is it can be used in the build logic.

Reviewer: ```suggestion
A task is *registered* in the build script using the `register()` method which allows it to then be used in the build logic.
```

.build.gradle.kts
[source,kotlin]
----
abstract class CreateFileTask : DefaultTask() {
    @TaskAction
    fun action() {
        val file = File("myfile.txt")
        file.createNewFile()
        file.writeText("HELLO FROM MY TASK")
    }
}

tasks.register<CreateFileTask>("createFileTask", ) {

}
----

== Task group and description

> Adding *group* and *description* statements to your tasks can help users understand how-to use your task:

Reviewer: ```suggestion
Setting the *group* and *description* properties on your tasks can help users understand how-to use your task:
```

[source]
----
------------------------------------------------------------
Tasks runnable from root project 'myTutorial'
------------------------------------------------------------

Build Setup tasks
-----------------
init - Initializes a new Gradle build.
wrapper - Generates Gradle wrapper files.

Help tasks
----------
buildEnvironment - Displays all buildscript dependencies declared in root project 'myTutorial'.
...
----

Once we apply a plugin to our project such as the `application` plugin, the default tasks available include:

[source]
----
Application tasks
-----------------
run - Runs this project as a JVM application

Build tasks
-----------
assemble - Assembles the outputs of this project.
build - Assembles and tests this project.

...

Documentation tasks
-------------------
javadoc - Generates Javadoc API documentation for the main source code.

...

Other tasks
-----------
compileJava - Compiles main Java source.

...
----

This is because the `application` plugin code contains additional tasks which are now available in your project.

== Task types

There are two types of tasks:

> 1. *Actionable tasks* have some action(s) attached to do work in your build: `compileJava`.

Reviewer: 💭 "Actionable" sounds like "able to be executed.  But lifecycle tasks can be executed too, they just do no work.  Is there a better word for these?
